frontend/src/lib/components/EditableText.svelte

```diff
@@ -0,0 +1,157 @@
+﻿<script lang="ts">
+	import { LoadingIcon } from '$lib/icons';
+	import { ZodString, z } from 'zod';
+	import { _changeProjectName } from '../../routes/project/[project_code]/+page';
+	import IconButton from './IconButton.svelte';
+	import { Form, lexSuperForm, lexSuperValidate } from '$lib/forms';
+
+	export let value: string | undefined | null = undefined;
+	export let disabled = false;
+	export let saveHandler: (newValue: string) => Promise<unknown>;
+	export let placeholder: string | undefined = undefined;
+	export let multiline = false;
+	export let validation: ZodString | undefined = undefined;
+
+	let initialValue: string | undefined | null;
+	let editing = false;
+	let saving = false;
+	const fieldId = 'id' + new Date().getTime();
+
+	let inputElem: HTMLInputElement | HTMLTextAreaElement;
+
+	const formSchema = z.object(validation ? { [fieldId]: validation } : {});
+	let { form, errors, valid, update, reset } = lexSuperForm(
+		formSchema,
+		{ taintedMessage: false },
+	);
+	$: error = $errors[fieldId]?.join(', ');
+
+	function startEditing() {
+		if (disabled) {
+			return;
+		}
+
+		initialValue = value;
+		form.set({[fieldId]: value ?? ''});
+		editing = true;
+	}
+
+	async function save() {
+		const newValue = inputElem.value;
+		if (newValue === initialValue) {
+			editing = false;
+			return;
+		}
+
+		saving = true;
+		editing = false;
+		await saveHandler(newValue);
```

Comment:
this could throw, so we might want to at the very least set `saving` to false in a finally. At best... I'm not sure, I think we need to setup some global error handlers to popup a toast for uncaught exceptions.

Reply:
something along these lines might be worth considering:  [next-app/src/lib/error/index.ts](https://github.com/sillsdev/web-languageforge/blob/develop/next-app/src/lib/error/index.ts)

Reply:
@longrunningprocess aha! thanks for that. I actually tried putting those on <svelte:window> which worked fairly well. The `onunhandledrejection ` always got called twice, which was a bit dissappointing. Perhaps your example will work better.